### ISSUE TITLE
Improve flexibility with constraining how venues are allocated

### ISSUE THREAD

**Issue by czlee** [czlee/tabbycat#318]
After #289 I've learnt that venue allocation can actually be quite complicated if done properly. Currently, it covers the ability to keep divisions in their assigned venue groups. Here are some other use cases:

* Constraining certain teams and/or adjudicators to rooms for accessibility reasons
* Having a preference for adjudication core and tab team members to be near the tab room

Note that the above includes _adjudicators_, which isn't currently possible in the current system, because venues are allocated before adjudicators are. Here's how I suggest that venue allocation should (eventually) work:

* Venue allocation should take place _after_ adjudicator allocations, _i.e._ it should be the last thing to happen before the draw is set in concrete.
* At the click of a button (or automatically), venues should just be allocated and, if input data is correct, there should ideally be no need to edit venues.
* Database models should include the following, all `ManyToManyField`s in the Venues app:
   * Team venue constraints, linking venue groups to teams, with priority (used for accessibility)
   * Adjudicator venue constraints, linking venue groups to adjudicators, with priority (used for accessibility/adj core)
   * Division venue constraints, linking venues groups to divisions, with priority. This replaces the current `venue_group` attribute of `Division`.
* The allocation procedure should go like this:
    1. For each pairing, determine which venue group it will go in. This would involve some resolution of potentially conflicting venue constraints. Priorities would be used to resolve conflicts, and if I'm really feeling ambitious I might do it so that the "set of venues" associated with a pairing doesn't necessarily match a venue group (say, it could be the intersection of two partially-conflicting groups). The user should be warned about any constraint that can't be met.
    2. Run the allocations procedure in [b78ba14](https://github.com/czlee/tabbycat/blob/b78ba1406d63c2698db2a3f996408a71baf2b939/draw/manager.py#L81-L100), which basically tries to satisfy all the pairing venue constraints, but in a way that doesn't just mean unconstrained pairings always get low-priority rooms.

Downsides:

* This adds an extra human step to draw generation, which for small tournaments without such complex requirements will be annoying. It might be worth extending our "round status" flow so that `CONFIRMED` refers to "pairings, adjudicators and venues all confirmed", as opposed to just pairings.
* With the flexibility I'm thinking of, it might be O(n²) or O(n³) or something, but I don't think in practice this will inflate the venue allocation time, since normally there are only a few constraints and they tend to play nicely.

Thoughts? Would this work okay with how division venue groups work in practice?

**Comment by czlee** [comment]
More precisely, I guess the best thing to do for each pairing is to find the subset of associated constraints with highest "order", subject to the consequent set of venues being nonempty, where by "order" I mean:

* Let _p_ be the constraint with highest priority in _A_ but not in _B_ (or −∞ if there is no such constraint), and let _q_ be the constraint with highest priority in _B_ but not in _A_ (or −∞). Then, if _p_ > _q_ then _A_ is higher than _B_, and if _q_ > _p_, then _B_ is higher than _A_.

So e.g. digits representing priorities of constraints: highest 4321, 432, 431, 43, 421, 42, 41, 4, 321, 32, 31, 3, 21, 2, 1 lowest

The algorithm for finding this is just to take the highest-priority constraint, and you keep adding the next constraint but only if it wouldn't make the set of allowable venues nonempty.

The other problem, of course, is that venue groups (or intersections thereof) might run out of venues. Ideally we'd want to prioritise those constraints with highest priority. I think we can do this by handling the pairings in descending order by the priority of the highest constraint of each pairing, removing venues from the pool when they're fully constrained, or something like that.

**Comment by philipbelesky** [comment]
#### UI Thoughts

It seems like constraints as a general feature would only be relevant for a handful of participants at large tournaments— say for the adj core and/or accessibility reasons. As such its fine adding them individually through the edit data UI or as part of a bulk import process. If for some reason the use case becomes more widespread then an easier way of editing or adding them would probably be needed.

#### WADL Related Thoughts

Divisions are assigned venue groups before rounds are generated and these are not changed once rounds have been generated. So I guess for the purposes of assigning venues to a Division's debates this would be modelled as each `Division` having a single constraint linking to a particular venue group? I think this would need a little bit of refactoring on my end — I think I access `division` through venue group on occasions. Although its usually accessed via `Debate` which should probably also be refactored to be a property only of the teams.  

I would also need to think about what overlap there is between this and the current WADL system for creating divisions and allocation teams to them. It's pretty messy at the moment. In theory it could work within the new system — `InstitutionalVenuePreferences` could just be bulk added as `TeamVenueConstraints` with relative priorities (and TeamVenuePreferences removed). 

However the `DivisionAllocator` process (a) happens before debates are drawn and (b) takes into account division size preferences quite strongly. So I don't think much of the codebase could be shared between this and the new system (in terms of creating allocations) but I could refactor `DivisionAllocator` to use the new constraints models and thus remove the existing preference models.

**Comment by czlee** [comment]
**UI:** I actually hadn't thought about the interface to add constraints. I agree that editing/adding can just be done in Jet.

**WADL:** First paragraph is correct.

We can have an `InstitutionalVenueConstraint` as well, so we don't have to do awkward bulk adding (it's easier to add teams and apply the same constraints this way). But I think this new paradigm is a lot more flexible than what WADL allows, and I'm not sure whether that is a good or bad thing for WADL. Assigning one venue per division is mutually exclusive with team/institution constraints (since the division would then absolutely constrain every venue). So if the reason why it's one division per venue group is external to team preference (_e.g._, you have a staff member on site to manage a division) then team/institution constraints don't make much sense for WADL. In that case, `DivisionAllocator` should just create a bunch of `DivisionVenueConstraints`, and as far as divisions are concerned nothing else has changed. On the other hand, if the purpose of WADL divisions is just to satisfy teams' preferences, then this new system would offer a lot more flexibility with how the draw works.

By the way, thinking about it now, it occurs to me that division allocation should be in its own app (in the Django sense, so in `divisions` rather than `tournaments`), since it's quite clearly a distinct function from everything else?

**Comment by philipbelesky** [comment]
Hmm, divisions being in its own app makes a degree of sense — in the end it only affects draw pairings, but on the way it touches on things like venue groups, institutions, etc. 

Not sure I completely follow your discussion of how this interacts with the WADL system. Is this thinking correct?

> Assigning one venue per division is mutually exclusive with team/institution constraints (since the division would then absolutely constrain every venue)

Quickly: you mean venuegroup, not venue right? I'm not sure I see what the problem is here per se? Is it that you end up with a system where WADL teams have TeamConstraints but also the divisions have DivisionContraints and these might create two constraints that can't be both satisfied? Except in WADL's case the DivisionConstraint's venuegroup will always be a subset of the TeamContraint's venuegroups as these tend to cover almost all options for each team.

> So if the reason why it's one division per venue group is external to team preference (e.g., you have a staff member on site to manage a division) then team/institution constraints don't make much sense for WADL.

A venue group would have multiple divisions that are constrained to it. These constraints in theory are related to team/institutional constraints even though (if as I proposed) they do so by way of a the `Division Allocation` using them to create DivisionConstraints. So I'm not sure they don't make sense for WADL; they are a model that can also express the current `Preferences` models, but do so in  adifferent manner. 

> In that case, DivisionAllocator should just create a bunch of DivisionVenueConstraints, and as far as divisions are concerned nothing else has changed. On the other hand, if the purpose of WADL divisions is just to satisfy teams' preferences, then this new system would offer a lot more flexibility with how the draw works.

That makes sense, but would the DivisionAllocator be informed by Constraints or by Preferences? The latter seems somewhat redundant with this system. 

Also if its not clear, `Venues` have no real bearing on WADL debates; they don't care about which teams get which rooms — `VenueGroups` are what is super important and their assignment precedes any draw generation.

**Comment by czlee** [comment]
> Quickly: you mean venuegroup, not venue right?

Yes, sorry.

> I'm not sure I see what the problem is here per se? ...

Yeah, you're completely right, in that case it wouldn't be harmful to have both, just redundant, so not mutually exclusive in practice like I claimed. I guess I was just thinking that assigning venues to meet constraints, and assigning divisions to meet preferences, are different concepts procedurally, so it would be confusing to overlay them. Maybe it won't be.

Also, I thought there might be a corner case where one team's preference might not be met during division allocations. Is there not?

That observation basically covers the rest of what you said. Preferences would be obsolete, like you say, and we'd just implement division constraints as acting on top of team constraints. If all division constraints are higher priority than all team constraints, and every division has only one venue group, then even in corner cases it'll assign divisions as you want them. Does that match/is that consistent with your thinking?

(When I said "one division per venue group" I actually meant one venue group per division, sorry, but that's immaterial now.)

**Comment by philipbelesky** [comment]
Yup, that all sounds good. In practice I don't think there are any corner cases where a team's preferences would not be met. This can be further ensured by adding to WADL's documentation that when adding constraints with the visual editor that the do so exhaustively; ie give a ranking to every venue group for every Institution,

**Comment by czlee** [comment]
It's probably worth making this explicit: I plan to write the algorithm so it can never break, _i.e._, no matter what constraints it is given, it will give _a_ solution of some sort, preferring a bad one to an incomplete one. My thinking is that venue allocations aren't really mission-critical, so it's better just to keep going than to insist on input perfection. It'll then feed warnings back to the UI if it couldn't satisfy someone. More precisely: it'll warn if someone didn't get _any_ of their constraints, since obviously if a team/adj/division/institution has multiple constraints, it makes no sense to satisfy all of them.

This means that if there are contradictory constraints, there will always be a warning. This includes the abovementioned corner case, where a team couldn't get one of its preferences in division allocation, causing the division constraint to contradict a team constraint. Would that bother you, _i.e._ should the warnings be suppressed when it's division vs team?

**Comment by philipbelesky** [comment]
No, I think having those warnings would be fine — the circumstance that would lead to a warning shouldn't have happened in the first place and if it does its a good way to highlight that fact.

**Comment by czlee** [comment]
A first implementation is up, but I'm a bit stuck on how to present the warning messages about constraints not being fulfilled. Any ideas?

**Comment by philipbelesky** [comment]
If I'm reading things right the messages would are shown on the draft draw page using the standard messaging system? I would have thought that incompatible constraints would be rare enough that simply stating "Debate of `X` vs `Y` could not be allocated venue because of incompatible constraints: `A`, `B`, `C`. Maybe also a suggestion to confirm the draw and then use Edit Venues to correct it?

**Comment by czlee** [comment]
Currently yeah, but it's far from clear to me that that's a sensible place to do it.

I'm a little confused about how your suggestion related to its being rare. Currently it just lists the unfulfilled constraints. The message that you suggest takes more work to construct and provides more useful information, right?

**Comment by philipbelesky** [comment]
Ah I was thinking of messages showing when there were mutually exclusive constraints rather than constraints being unfulfilled. The later should still be relatively rare if the use case is 'adj core plus a couple of specific teams' though, at which point a message or two wouldn't be overwhelming?

If there were situations were there would be say 4+ messages being displayed it would be nice for them to be displayed in the table itself as a notes. But I don't think that's required for the feature to be implemented, just a nice to have for edge cases where constraints are heavily used.

**Comment by czlee** [comment]
Hmm, we could do that. It wouldn't be too hard, and would have the advantage of being persistent, unlike messages. More specifically, I'd have a new model/table `VenueConstraintFailure` or something like that in the venues app, with a `ForeignKey` to `Debate`, then relevant views can find them if needed.

**Comment by czlee** [comment]
Okay, I realized that we can just figure out the unmet constraints directly from the constraints and draw, and that that's probably a better idea than storing them during auto-allocate, since then they would also be calculated for human-allocated venues. Also, they can be displayed in the "conflicts" column, which they now are. So I did that.

Also, I moved the venue-specific views to the venues app and converted them to class-based views.

I also added an "auto-allocate" button to the "edit venues" view, but it's not pretty. It just auto-allocates venues then reloads the page (redirects to itself). It should probably use an ajax request like the adjudicator auto-allocate does but I don't have enough time/am too lazy to figure out how to do it. Do you want to, @philipbelesky? It's functional as it is, just not best practice UI-wise I guess.

If you're happy with it how it stands now then it's mergeable to develop, but holding off in case you want to fix my auto-allocate kludge.

**Comment by philipbelesky** [comment]
Wooo! I'd just go ahead and merge it. Edit venues is pretty niche, and I'm hopefully redeveloping all the drag/drop views over the next month anyway, so I can setup the proper ajax method when that happens.
